File: node_server/sim/buildings/modules/simModule.js
import { Simulation } from '../../simulation.js';

export class SimModule {

  // TODO: refactor that either modules attach their state lazily to buildings
  // or that buildings define state and we use that existence to run the module

  /**
   * @param {World} world
   * @param {Building} building
   */
  simulate(world, building) {
    // Implement in subclass
  }

  /**
   * @param {World} world
   * @param {Building} building
   */
  dispose(world, building) {
    // default does nothing
  }
}

File: CitySim.Simulation/Systems/Buildings/Modules/SimModule.cs
using CitySim.Simulation.Model.Buildings;

namespace CitySim.Simulation.Systems.Buildings.Modules;

using CitySim.Simulation.Model;

public abstract class SimModule
{
    public abstract void Simulate(World world, Building building);
    public virtual  void Dispose (World world, Building building) { }
}
